raft: make sure to retain the existing voters including the current leader (topology coordinator)

emdroid opened the pull request:
Fix an issue in the voter calculator where existing voters were not retained across data centers and racks in certain scenarios. This occurred when voters were distributed across more data centers and racks than the maximum allowed number of voters.
Previously, the prioritization logic for data centers and racks did not consider the number of existing assigned voters. It only prioritized nodes within a single data center or rack, which could result in unnecessary reassignment of voters.
Improved the prioritization logic to account for the number of existing assigned voters in each data center and rack.

Additionally, the limited voters feature did not account for the existing topology coordinator (Raft leader) when selecting voters to be removed. As a result, the limited voters calculator could inadvertently remove the votership of the topology coordinator, triggering unnecessary Raft leader re-election.
To address this, the topology coordinator's votership status is now preserved unless absolutely necessary. When choosing between otherwise equivalent voters, the node other than the existing topology coordinator is prioritized for removal.

This change ensures a more stable voter distribution and reduces unnecessary voter reassignments.

The limited voters calculator is refactored to use a priority queue for sorting nodes by their priorities. This change simplifies the voter selection logic and makes it more extensible for future enhancements, such as supporting more complex priority calculations.

Fixes: scylladb/scylladb#23950
Fixes: scylladb/scylladb#23588
Fixes: scylladb/scylladb#23786

No backport: The limited voters feature is currently only present in master.

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -398,15 +416,25 @@ future<> group0_voter_handler::update_nodes(
+
+    if (!raft_server.is_leader()) {
+        // Only expected to be called by the topology coordinator
+        on_internal_error(rvlogger, "update_nodes called on non-leader node");
```
The return shouldn't be needed - as the `on_internal_error()` doc says:
```cpp
/// Depending on the value passed to set_abort_on_internal_error, this
/// will either abort or throw a std::runtime_error.
```

emdroid commented:
Verified that it fixes the #23786:
* before this change: 2 of 100 runs failed
* with this change: 0 of 500 runs failed

emdroid commented:
New version:
* account for the possible race when the topology coordinator is losing the leadership (print just a warning instead of the hard error)

[diff](https://github.com/scylladb/scylladb/compare/abfa7bbd0dbb066bd2cbad5cd40ebbadddc2ee6b..85976bd282cb7bb6d2a38b260fbadfdac0d157b2)

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -24,16 +23,41 @@ namespace {
+            priority += node_priority::voter;
+        }
+        if (node.flags.is_leader) {
+            priority += node_priority::leader;
```
No, increased priority is for selecting the voters, so higher priority means less chance of being removed from voters.

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -398,15 +416,26 @@ future<> group0_voter_handler::update_nodes(
     group0_voter_calculator::nodes_list_t nodes;
 
-    auto add_nodes = [this, &nodes, &group0_config](const std::set<locator::host_id>& nodes_set, bool is_alive) {
+    auto add_nodes = [this, &nodes, &group0_config, &leader_id](const std::set<locator::host_id>& nodes_set, bool is_alive) {
```
Added a helper to avoid unnecessary duplication

bitpathfinder commented on service/raft/group0_voter_handler.cc:
```diff
@@ -47,22 +63,46 @@ bool operator<(const is_voter_t& lhs, const is_voter_t& rhs) {
 
-    static nodes_map_t create_nodes_map(std::ranges::input_range auto&& nodes) {
-        return nodes | std::views::transform([](const auto& node_entry) {
+    static nodes_store_t create_nodes_list(std::ranges::input_range auto&& nodes, size_t* alive_nodes_remaining, size_t* dead_voters_remaining) {
```
Why use pointers when they cannot be null? imho better to use references in these cases.

emdroid replied:
the reason is to make it more obvious at the call site that it is an output parameter and the value is being modified - in particular because it is used for modifying class attributes in the constructor initializer list

bitpathfinder replied:
By adding a pointer to the function interface you are creating more unnecessary invariants in the function. Besides, imho nobody should assume that parameters are not changed in the function.
Maybe consider making the function a non static and access these members without passing them to the function?

bitpathfinder replied:
Well if you want it to be explicit, then probably you could just paste the function contents to the constructor (because in essence it is a constructor)

emdroid replied:
replaced by references

gusev-p commented on service/raft/group0_voter_handler.cc:
```diff
@@ -56,7 +56,7 @@ class rack_info {
         return nodes | std::views::transform([](const auto& node_entry) {
             const auto& [id, node] = node_entry;
-            return std::make_pair(std::make_tuple(is_alive_t{node.is_alive}, is_voter_t{node.is_voter}), id);
+            return std::make_pair(std::make_tuple(is_alive_t{node.get().is_alive}, is_voter_t{node.get().is_voter}), id);
```
The code in `group0_voter_calculator` seems more complicated than needed. We have at most 5 voters - we can linearly scan all nodes for max/min five times - no need for separate classes for rack/dc or priority_queues.

emdroid replied:
well, the code was originally written when it wasn't decided yet how many max voters we'll want to have (and even the adaptive maximum based on DC count was considered), so the code is fairly generic to support arbitrary number of max voters ... we can think of simplifying this, but I would keep that separate from this change

gusev-p commented on service/raft/group0_voter_handler.cc:
```diff
@@ -56,7 +56,7 @@ class rack_info {
-            return std::make_pair(std::make_tuple(is_alive_t{node.is_alive}, is_voter_t{node.is_voter}), id);
+            return std::make_pair(std::make_tuple(is_alive_t{node.get().is_alive}, is_voter_t{node.get().is_voter}), id);
         }) | std::ranges::to<nodes_map_t>();
     }
```
It's quite difficult to navigate through the code here because of too heavy use of ranges and auto in func signatures. Sometimes I need to make several jumps between implementation pieces to understand what is mapped to what, e.g. in "explicit datacenter_info(std::ranges::sized_range auto&& nodes)".

emdroid replied:
yeah ... these (like the `std::ranges::sized_range auto&& nodes`) in particular are used to not have to construct a new temporary structure (like a map) from the range, but to use the range views directly (i.e. kind of performance reasons)

gusev-p commented on service/raft/group0_voter_calculator.hh:
```diff
@@ -36,11 +36,19 @@ public:
+    // Represents the flags associated with a node in the group 0.
+    //
+    // These flags provide additional information about the node's status and role.
+    struct node_flags {
```
why separate class node_flags? can't we put another bool right into node_descriptor?

emdroid replied:
yes, removed and put the bool directly to the node descriptor ... initially my idea was to maybe look into using bitfield flags, but looked at the other parts of the code that mostly use this kind of flag classes. But then it indeed doesn't make much sense to have a separate flags struct (as the `is_alive`, `is_voter` are in fact also kind of flags)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16979/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16979/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16979/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/group0_voter_calculator_test
:small_blue_diamond: cluster/test_raft_voters
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4938/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8110/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16179/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16979/)
#### Build Details:
- Duration: 7 hr 8 min
- Builder: spider5.cloudius-systems.com

emdroid commented:
New version:
* removed separate flags
* introduced named type for the priority value
* further code simplifications

[diff](https://github.com/scylladb/scylladb/compare/2b4913ce553ecd158ee77dc2eba5df151b224d84..bb43a1d4bb8549c74cdfbf3a7a015ba894990243)

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -161,21 +161,16 @@ class datacenter_info {
+            nodes_by_rack[node.get().rack].emplace_back(id, node);
         }
 
+        const auto racks = nodes_by_rack | std::views::keys | std::ranges::views::transform([&nodes_by_rack](const auto& rack) {
```
changed to just return (added the `std::ranges::to`)

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -306,13 +301,11 @@ class calculator_impl {
-            }
-            return nodes_by_dc;
-        });
+        using dc_nodes_t = std::vector<std::pair<raft::server_id, std::reference_wrapper<const group0_voter_calculator::node_descriptor>>>;
```
right, added a single named type for these

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -24,16 +23,33 @@ namespace {
+
+    static constexpr int calculate(const group0_voter_calculator::node_descriptor& node) {
+        int priority = 0;
+        if (node.is_alive) {
```
I'd like to keep the code more clear and readable (I don't expect this part to be a performance bottleneck) - another advantage of the current code is that the git diff will be cleaner when adding flags (a.g. for the zero token nodes)

bitpathfinder replied:
ok

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -24,16 +23,33 @@ namespace {
+    // It should be smaller than the alive node priority (to still prefer alive nodes to dead voters).
+    static constexpr int voter = 1;
+
+    static constexpr int calculate(const group0_voter_calculator::node_descriptor& node) {
```
makes sense, added a typedef

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -47,22 +63,46 @@ bool operator<(const is_voter_t& lhs, const is_voter_t& rhs) {
+    size_t _alive_nodes_remaining = 0;
+    size_t _dead_voters_remaining = 0;
+
+    using node_info_t = std::tuple<int, raft::server_id,
```
yes, using the new typedef here now for clarity

emdroid commented:
New version:
* replace the pointer output parameters by reference

[diff](https://github.com/scylladb/scylladb/compare/bb43a1d4bb8549c74cdfbf3a7a015ba894990243..5189ee7ee4b76321656fb216f9f6572940c0d5db)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16998/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16998/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16998/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/group0_voter_calculator_test
:small_blue_diamond: cluster/test_raft_voters
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4950/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16191/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8123/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16998/)
#### Build Details:
- Duration: 7 hr 14 min
- Builder: spider3.cloudius-systems.com

bitpathfinder commented on service/raft/group0_voter_handler.cc:
```diff
@@ -297,18 +297,17 @@ class calculator_impl {
 
-        return nodes_by_dc | std::views::keys | std::ranges::views::transform([&nodes_by_dc, largest_dc_size](const auto& dc) {
+        return nodes_by_dc | std::views::keys | std::ranges::views::transform([&nodes_by_dc, &largest_dc_size](const auto& dc) {
             const auto& dc_nodes = nodes_by_dc.at(dc);
```
it looks a bit strange - first to filter the keys, but then to capture the `nodes_by_dc` to get the value. why don't just 
```C++ 
return nodes_by_dc | std::ranges::views::transform([largest_dc_size](const auto& item) {
   const auto& dc_nodes = item.second;
   largest_dc_size = std::max<size_t>(&largest_dc_size, dc_nodes.size());
   return dc_nodes;
}) | std::ranges::to<datacenters_store_t>();

emdroid replied:
This was in fact the relict of the previous way of retrieving the nodes_by_dc into an `std::unordered_multimap` where the keys needed to be converted to a set first. Here in fact we even don't need `std::ranges::views::transform` at all and can use `std::views::values` directly, updated (the `largest_dc_size` calculation can be factored out of the loop and use `max_element` for it separately without the need of transform)

bitpathfinder commented on service/raft/group0_voter_handler.cc:
```diff
@@ -24,16 +23,35 @@ namespace {
-bool operator<(const is_alive_t& lhs, const is_alive_t& rhs) {
-    return static_cast<bool>(lhs) < static_cast<bool>(rhs);
-}
+    using value_t = int;
```
Can the value be negative?

emdroid commented:
New version:
* refs to iterators not necessary
* directly use `nodes_by_X | std::views::keys` instead of `nodes_by_X | std::views::keys | std::ranges::views::transform(... to values ...)`
* refactor the `largest_dc_size` calculation (use `max_element`)

[diff](https://github.com/scylladb/scylladb/compare/5e9693c163fa41291e3a27d6ae922e90ca27b1f1..a005d6c93b4d16dd4818223cfceb32d89c9c52ab)

emdroid commented:
^ rebased ^

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -396,7 +396,19 @@ future<> group0_voter_handler::update_nodes(
+    auto add_nodes_to_list = [this, &add_node](const std::set<locator::host_id>& nodes_set, bool is_alive) {
         for (const auto& host_id : nodes_set) {
             const raft::server_id id{host_id.uuid()};
             const auto& it = _topology.normal_nodes.find(id);
```
this was coming from the initial limited voters PR, but it's true that the iterators don't need a reference (in this case even a pinning ref), changed

bitpathfinder replied:
Yeah, better to not make references to temporary objects, even tho in this case the lifetime of the object is extended.

patjed41 commented on test/cluster/test_raft_voters.py:
```diff
@@ -176,3 +176,60 @@ async def upgrade_server(srv):
+
+
+@pytest.mark.asyncio
+async def test_raft_limited_voters_retain_coordinator(manager: ManagerClient):
```
Don't `test_raft_limited_voters_retain_coordinator` and `leader_is_retained_as_voter` test the same exact thing?

emdroid replied:
They basically are, only that the python test is also testing for the code setting the topology coordinator flag.

patjed41 replied:
Let's keep both tests.

We discussed the idea where the boost test could check two cases: when `ids[0]` is the leader and `ids[1]` is the leader. Currently, it could theoretically happen that the test always passes, but the implementation is incorrect. For example, `ids[0]` is the leader in the test, but the implementation could ignore leadership and just make the last node in a DC a non-voter, and the test would pass. It's not the case now, but the implementation could change, so we better not rely on it.

If we change the boost test this way, then both these tests will have a purpose. I would change only the boost test, since it takes negligible running time.

patjed41 commented on service/raft/group0_voter_handler.cc:
```diff
@@ -41,6 +41,11 @@ struct node_priority {
     // It should be smaller than the alive node priority (to still prefer alive nodes to dead voters).
     static constexpr value_t voter = 1;
 
```
Do I understand correctly that leadership is considered only on the rack level? When we chose the next DC, a DC with the leader can have the same priority or lower priority than a DC without the leader, right? Why are we sure that we won't make the leader non-voter unless absolutely necessary? I don't have a solid argument, but I also don't have a counterexample.

emdroid replied:
Added the tests and fixes for also prioritizing the data centers correctly (which indeed wasn't the case).

This also led to realizing another bug: #23950 where the same issue is actually there for retaining the existing voters - these also need to be considered for the prioritization of data centers and racks (whereas they were only considered on the rack nodes level).

patjed41 commented on service/raft/group0_voter_handler.cc:
```diff
@@ -404,22 +412,35 @@ future<> group0_voter_handler::update_nodes(
+
+    const auto& leader_id = raft_server.current_leader();
+
+    // There may be a race with the topology coordinator losing the leadership.
+    // The state will be updated once the new leader is elected.
+    if (!leader_id) {
+        rvlogger.warn("No leader, skipping voter calculation");
+        co_return;
+    }
```
> The important question is what happens if the local node does lose leadership and changing the voters still succeeds. I think it's possible because `raft::server::config::enable_forwarding == true` for group 0.

Maybe it is impossible because of `topology_coordinator::_as` that is triggered when the node loses leadership.

emdroid replied:
Yes, the idea was to just be sure we have a valid `leader_id` (in a previous PR iteration there was a stricter check, but in fact I have sen it failing a test). Perhaps we can do the check just before setting the voters (so there shouldn't come any preemption point in between), but I'm still not sure if it really fully resolves the race.

emdroid replied:
Removed the check... in the case the leader id will be empty (the case of switching the topology coordinator), no voter will be marked as the leader, but this will not break the calculator itself, and then we will depend on the abort source cancelling the members modification.

patjed41 replied:
> and then we will depend on the abort source cancelling the members modification

It would be best to investigate whether we have this guarantee. It shouldn't be hard but tedious, I imagine a test with a combination of sleeps, error injections, and log browsing. We have more important stuff to do, though, so let's not do it now and get this PR merged quickly. Maybe we can get back to it in the future.

Also, even if the old leader caused the new leader to lose leadership, this situation shouldn't repeat indefinitely. We should eventually select a leader that won't be distracted. It should also happen in the following leader election.

bhalevy commented:
@tgrabiec can you please review too?

emdroid commented:
New version:
* test the voter calculation with various initial leader nodes

[diff](https://github.com/scylladb/scylladb/compare/9688de0f041f92dd823fc7e507b15223d93cd4ed..f81cafef80d7bcf6e4d53e96a1f9f00069f5cc62)

emdroid commented:
New version:
* consider the ownership of the leader (topology coordinator) on the data center and rack level prioritization (added an unit test)
* fix the issue #23950 where the existing voters might actually not be retained between data centers and racks under some circumstances - need to consider the number of existing voters also on the level of prioritization of data centers and racks

[diff](https://github.com/scylladb/scylladb/compare/f81cafef80d7bcf6e4d53e96a1f9f00069f5cc62..44510c64d181ff4aca9565d9439b8f4fb6f524ce)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17032/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17032/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17032/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/group0_voter_calculator_test
:small_blue_diamond: cluster/test_raft_voters
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16221/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4980/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8153/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17032/)
#### Build Details:
- Duration: 7 hr 22 min
- Builder: spider6.cloudius-systems.com

patjed41 commented on test/boost/group0_voter_calculator_test.cc:
```diff
@@ -869,4 +869,74 @@ BOOST_AUTO_TEST_CASE(dcs_preferred_over_racks) {
+
+    const service::group0_voter_calculator voter_calc{max_voters};
+
+    // Act: Add the nodes (one of them is already a voter).
```
There are 3 voters, not 1. The same in the test case below.

emdroid replied:
fixed

patjed41 commented on service/raft/group0_voter_handler.cc:
```diff
@@ -169,12 +187,17 @@ class rack_info {
+        // Third criteria: The ownership of the leader node
+        if (rack1._owns_leader != rack2._owns_leader) {
+            return rack2._owns_leader;
+        }
```
I'm not sure if this should be the third criteria or the second criteria. A rack with the leader will have lower priority than a rack with two voters, which is suspicious. It may be correct, perhaps we have a guarantee that we will later take the leader as a voter anyway. However, it would be better to not worry about this and take the leader first if it is alive. So, maybe we should introduce a second criteria "the ownership of the live leader" instead?

The same goes for `datacenter_info`.

emdroid replied:
Right, reordered to be second criteria (before the number of existing voters)

patjed41 commented on service/raft/group0_voter_handler.cc:
```diff
@@ -197,19 +220,23 @@ class datacenter_info {
-                }
+            if (node.get().is_alive && node.get().is_voter) {
+                ++alive_voters_remaining;
+            }
```
You should have
```cpp
if (node.get().is_alive && node.get().is_voter) {
    ++alive_voters_remaining;
}
```
 in the second last commit to avoid this diff in the last commit.

emdroid replied:
fixed

patjed41 commented on test/boost/group0_voter_calculator_test.cc:
```diff
@@ -939,4 +940,105 @@ BOOST_AUTO_TEST_CASE(existing_voters_are_kept_across_racks) {
 }
 
 
+BOOST_DATA_TEST_CASE(leader_is_retained_as_voter, boost::unit_test::data::make({0, 1, 2}), leader_node_idx) {
```
Nice, I didn't know we can parameterize boost tests this way.

patjed41 commented on test/boost/group0_voter_calculator_test.cc:
```diff
@@ -939,4 +940,105 @@ BOOST_AUTO_TEST_CASE(existing_voters_are_kept_across_racks) {
+
+    // Assert: The current leader is retained as a voter.
+
+    BOOST_CHECK_GE(voters.size(), 1);
```
If `voters.size() == 2`, this test doesn't make sense. It will always trivially pass because all nodes are voters. I think we should check that `voters.size() == 1`.

emdroid replied:
I added a comment about it, basically currently we will get 2 because of the condition in the handler that allows for any number of voters below 3. This will be addressed in #23266 later, but I didn't want to remove the condition right now, as it has a potential to break some tests, so I'll do it separately.

patjed41 reviewed (changes requested):
Looks good, some small comments and questions.

You could update the PR title, the cover letter, and the commit messages a bit after fixing #23950. I would argue that the fix for #23950 is the main part of this PR. The leader can be simply considered as a special case of a voter with the highest priority and #23588 can be considered as an adjustment to #23950. I won't insist on doing this boring stuff, though. The current version is good enough.

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -56,6 +56,9 @@ struct node_priority {
 using nodes_ref_list_t = std::vector<std::pair<raft::server_id, std::reference_wrapper<const group0_voter_calculator::node_descriptor>>>;
 
+using select_voter_result_t = std::tuple<raft::server_id, std::reference_wrapper<const group0_voter_calculator::node_descriptor>>;
+using select_voter_result = std::optional<select_voter_result_t>;
```
Removed both, as I reverted the `select_voter_result` changes so these are no longer necessary.

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -88,14 +92,20 @@ class rack_info {
+                    ++alive_voters_remaining;
+                } else {
+                    ++alive_nodes_remaining;
+                }
```
Right ... it should work either way, but for less confusion the variable is now being updated in either case.

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -149,15 +164,18 @@ class rack_info {
 
-        // Second criteria: The number of alive nodes (voters and non-voters) remaining (higher has more priority)
+        // Second criteria: The number of alive voters remaining (higher has more priority)
+        if (rack1._alive_voters_remaining != rack2._alive_voters_remaining) {
```
Good point, switched the order of criteria ... I initially didn't have resetting the flag back so it would always prefer the leader holding dc/rack, but now the code resets the flag once we process the leader, so that is no longer an issue and it's less confusing.

patjed41 replied:
I think you didn't understand what I meant. I was talking about the first criteria being about different voters than other criteria. The first criteria is about voters already assigned in the config we are building right now, but the other criteria are about voters/nodes from the old config that are not yet assigned as voters in the new config.

The comments
> // First criteria: The number of already assigned voters (lower has more priority)

> // Second criteria: The number of alive voters remaining (higher has more priority)

mention some voters, but they don't specify where exactly these voters are "already assigned" or "remaining".  It's easy to misunderstand these comments unless you know the algorithm well. There are also variables like `_voters_count` and `_alive_voters_remaining` that are related to completely different voters, but it's not clear from their names.

One idea I have is to define somewhere "old voters/nodes" (from the old config) and "new voters/nodes" (from the config being built) and consistently use these names everywhere (comments, variable names, etc.). This would arguably make the code easier to understand.  We could do something like this, but it requires some work and is not very important, for sure out of the scope of this PR.

emdroid replied:
Updated for more clarity - referring to the "old" voters as "existing" voters and the new voters as (newly) "assigned" voters.

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -149,15 +164,18 @@ class rack_info {
+            return rack1._alive_voters_remaining < rack2._alive_voters_remaining;
+        }
 
+        // Third criteria: The number of alive nodes (voters and non-voters) remaining (higher has more priority)
```
As above, keeping the `_alive_nodes_remaining` and counting both cases.

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -201,29 +228,36 @@ class datacenter_info {
 
+            const auto& [voter_id, node] = *voter_res;
+
+            if (node.get().is_alive && node.get().is_voter) {
```
Makes sense ... one slight issue is that we remove the node information (we only hold it in the priority queue) once we dequeue it, so it is no longer available in the struct when we get the result.
But added a pointer that will keep note of the currently selected voter node information so that we can access it for checking the voter attributes.

patjed41 replied:
> Makes sense ... one slight issue is that we remove the node information (we only hold it in the priority queue) once we dequeue it, so it is no longer available in the struct when we get the result.

My bad, I forgot about it when writing this comment.

Right now I'm not sure which version is better, the current one or the previous one. I think we can just keep the current one.

Another possible solution is to refactor `rack_info` so that the priority queue contains only `raft::server_id`s and there is also a map with the `raft::server_id`->`node_descriptor` mapping. Then, we would remove only `raft::server_id`s from the priority queue. I'm not sure if it's better, but even if it is, it's probably not worth the effort.

emdroid commented on test/boost/group0_voter_calculator_test.cc:
```diff
@@ -869,4 +869,74 @@ BOOST_AUTO_TEST_CASE(dcs_preferred_over_racks) {
+            {ids[1], {.datacenter = "dc-2", .rack = "rack", .is_voter = false, .is_alive = true}},
+            {ids[2], {.datacenter = "dc-3", .rack = "rack", .is_voter = true, .is_alive = true}},
+            {ids[3], {.datacenter = "dc-4", .rack = "rack", .is_voter = false, .is_alive = true}},
+            {ids[4], {.datacenter = "dc-4", .rack = "rack", .is_voter = true, .is_alive = true}},
```
good catch, fixed

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17052/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17052/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17052/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/group0_voter_calculator_test
:small_blue_diamond: cluster/test_raft_voters
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5001/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8172/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16240/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17052/)
#### Build Details:
- Duration: 9 hr 20 min
- Builder: spider8.cloudius-systems.com

patjed41 reviewed (commented):
Looks good, only two nits

emdroid commented on service/raft/group0_voter_handler.cc:
```diff
@@ -36,6 +36,9 @@ bool operator<(const is_voter_t& lhs, const is_voter_t& rhs) {
 }
```
fixed

bitpathfinder commented on service/raft/group0_voter_handler.cc:
```diff
@@ -306,21 +299,18 @@ class calculator_impl {
-            *largest_dc_size = std::max<size_t>(*largest_dc_size, dc_nodes.size());
-            return dc_nodes;
-        });
+        if (!nodes_by_dc.empty()) {
```
nit: Finding max could be done also in the cycle above.

emdroid replied:
It could, however it would need to be checked at every node (at every emplace of the node per DC). The `std::max_element()` done afterwards makes it more efficient, only comparing the final sizes (although there is most likely no significant performance impact of either way).

bitpathfinder commented on service/raft/group0_voter_handler.cc:
```diff
@@ -396,28 +396,34 @@ future<> group0_voter_handler::update_nodes(
 
-    add_nodes(nodes_alive, true);
-    add_nodes(nodes_dead, false);
+    add_nodes_to_list(nodes_alive, true);
```
nit: imho the name is a bit confusing. It's not clear that the nodes are added *from* the provided list, not to the list. What do you think about naming it `add_nodes_from_list`?

emdroid replied:
Right, renamed. The "to_list" referred to the final complete list where the nodes are being added to, but you're right that when the list to add from is being passed in, the "add_from_list" naming is more appropriate.

bitpathfinder commented on service/raft/group0_voter_handler.cc:
```diff
@@ -427,20 +433,12 @@ future<> group0_voter_handler::update_nodes(
-        const auto& it = _topology.normal_nodes.find(id);
+        const auto it = _topology.normal_nodes.find(id);
         if (it == _topology.normal_nodes.end()) {
             rvlogger.warn("Node {} to be added not found in the topology - can't add as a voter candidate", id);
```
nit: maybe '"Node {} was not found in the topology, can't add as a voter candidate"' otherwise, message states 2 times that the node is to be added.

bitpathfinder commented on service/raft/group0_voter_handler.cc:
```diff
@@ -125,16 +137,29 @@ class rack_info {
         }
 
         ++_voters_count;
+        _selected_voter = &node.get();
```
How's lifetime of the `node` managed?

emdroid replied:
The references are to the node instances in the list `nodes_filtered` created inside of `group0_voter_calculator::distribute_voters()`, so they should be valid for the whole time the calculation is being done.

I'm actually about to refactor this a bit as we don't even need to use the references at all - the input parameters of both `nodes` and `nodes_filtered` are maps of `raft::server_id` to `node_descriptor`, so we can just keep working with the raft id, drom all the node descriptor references and do a lookup in the original map every time we need any info about the node. This should simplify the code quite a lot.

But as it is a lateral change to this fix PR, I'll do it as a separate follow-up.

bitpathfinder commented on service/raft/group0_voter_handler.cc:
```diff
@@ -177,13 +205,22 @@ class datacenter_info {
         std::unordered_map<std::string_view, nodes_ref_list_t> nodes_by_rack;
         for (const auto& [id, node] : nodes) {
             nodes_by_rack[node.get().rack].emplace_back(id, node);
+            if (node.get().is_alive) {
```
Why not `if (node.get().is_alive && node.get().is_voter) {++alive_voters_remaining;}`

emdroid replied:
It's to have a smaller diff for the change to consider the leader (which is also only checked for the alive nodes, so it makes sense to split the condition then - so the condition is left split for the other change as well).

bitpathfinder commented on service/raft/group0_voter_handler.cc:
```diff
@@ -218,6 +255,15 @@ class datacenter_info {
 
+            const auto& node = rack.get_selected_voter_info();
+
+            if (node.is_alive) {
```
Why not `if (node.get().is_alive && node.get().is_voter) {SCYLLA_ASSERT(_alive_voters_remaining > 0); --alive_voters_remaining;}`

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17118/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17118/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17118/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/group0_voter_calculator_test
:small_blue_diamond: cluster/test_raft_voters
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5058/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16303/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8232/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17118/)
#### Build Details:
- Duration: 8 hr 45 min
- Builder: spider8.cloudius-systems.com

patjed41 commented:
queued
